### DISCUSSION TITLE
How to escape question mark at the beginning of the password? (Gsec)

### DISCUSSION THREAD

**Discussion by Prochy** [FirebirdSQL/firebird#8308] [Q&A]
I was not able to found in the documentation if it is somehow limited using question mark at the beginning of the password. I found this FAQ and there is written there is no character limitation. [FAQ](https://firebirdfaq.org/faq265/)

When I try executing following command: 
`gsec -user SYSDBA -password "masterkey" -modify SYSDBA -pw "??"`
The command failed and instead of any error help is shown.
Theoretically if I were successful and then I would try changing it back
`gsec -user SYSDBA -password "??" -modify SYSDBA -pw "masterkey"`
I got "**operation already specified**. 

Not sure if it's not supported or I have to somehow escape it. I didn't find any mention in the documentation. 
Note: The problem is only with the question mark at the beginning... -pw "w????" this works fine

**Comment by hvlad** [comment]
The problem is in command processor, not in gsec. Try single quotes.
Also, note, gsec is deprecated, better use isql and manage users via SQL.

**Comment by aafemt** [comment]
According to the documentation even bash doesn't expand globs in quoted parameters and otherwise question mark is not a special symbol. I wonder what command processor is used?

**Reply by mrotteveel** [reply]
Given I can reproduce it in Windows Command Prompt, I guess Windows Command Prompt :) Using single quotes as suggested by Vlad works.

**Reply by mrotteveel** [reply]
I spoke too soon, using single quotes in Command Prompt will make the password literally `'??'` (thus the quotes are part of the password).

**Reply by aafemt** [reply]
Looking at the sources of gsec it is platform-independent and definitely a bug. I see no way to escape or workaround it.

**Reply by mrotteveel** [reply]
@aafemt I guess you mean in `get_line`, the 
```c++
if (first)
{
	first = false;
	if (c != '?')
	{
		*cursor++ = '-';
		count--;
	}
}
```

**Reply by aafemt** [reply]
No, relevant piece of code is this (gsec.cpp:839):
```
		if (*string == '?' || strcmp(string, "-?") == 0)
			user_data->op = HELP_OPER;
```
1. It checks only the first symbol, i.e. "?anything" is recognized as a help switch.
2. This condition is the first in chain, i.e. it disregard position of this string.

**Comment by hvlad** [comment]
Sorry, I was wrong. It is definitely bug in gsec's `get_switches()`.

**Comment by Prochy** [comment]
I will put question in separate thread so it's more visible. 

Thanks guys for analysing the issue. Is it worth creating a bug for gsec when it's deprecated? I guess it's low priority. Or at least create a bug so it is recorded when someone will be looking for the same issue?

**Reply by hvlad** [reply]
> Is it worth creating a bug for gsec when it's deprecated? I guess it's low priority. Or at least create a bug so it is recorded when someone will be looking for the same issue?

Bug is a bug, it is better to create a ticket for it, imho. No promise to fix it, though ;)
